src/EditorFeatures/CSharp/ConvertNamespace/ConvertNamespaceCommandHandler.cs

```diff
@@ -0,0 +1,168 @@
+        /// value.
+        /// </summary>
+        private static readonly OptionSet s_optionSet = new OptionValueSet(
+            ImmutableDictionary<OptionKey, object?>.Empty.Add(
+                new OptionKey(CSharpCodeStyleOptions.NamespaceDeclarations.ToPublicOption()),
+                new CodeStyleOption2<NamespaceDeclarationPreference>(
+                    NamespaceDeclarationPreference.FileScoped,
+                    NotificationOption2.Suggestion)));
+
+        private readonly ITextUndoHistoryRegistry _textUndoHistoryRegistry;
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public ConvertNamespaceCommandHandler(
+            ITextUndoHistoryRegistry textUndoHistoryRegistry,
+            IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            _textUndoHistoryRegistry = textUndoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public CommandState GetCommandState(TypeCharCommandArgs args, Func<CommandState> nextCommandHandler)
+            => nextCommandHandler();
+
+        public string DisplayName => CSharpAnalyzersResources.Convert_to_file_scoped_namespace;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            // Attempt to convert the block-namespace to a file-scoped namespace if we're at the right location.
+            var convertedRoot = ConvertNamespaceCommandHandler.ConvertNamespace(args, executionContext);
+
+            // No matter if we succeeded or not, insert the semicolon.  This way, when we convert, the user can still
+            // hit ctrl-z to get back to the code with just the semicolon inserted.
+            nextCommandHandler();
+
+            // If we weren't on a block namespace (or couldn't convert it for some reason), then bail out after
+            // inserting the semicolon.
+            if (convertedRoot == null)
+                return;
+
+            // Otherwise, make a transaction for the edit and replace the buffer with the final text.
+            using var transaction = CaretPreservingEditTransaction.TryCreate(
+                this.DisplayName, args.TextView, _textUndoHistoryRegistry, _editorOperationsFactoryService);
+
+            var edit = args.SubjectBuffer.CreateEdit(EditOptions.DefaultMinimalChange, reiteratedVersionNumber: null, editTag: null);
+            edit.Replace(new Span(0, args.SubjectBuffer.CurrentSnapshot.Length), convertedRoot.ToFullString());
+
+            edit.Apply();
+
+            // Attempt to place the caret right after the semicolon of the file-scoped namespace.
+            var annotatedToken = convertedRoot.GetAnnotatedTokens(s_annotation).FirstOrDefault();
+            if (annotatedToken != default)
```

Comment:
In general, i don't think this should be gated.  Realistically we have to ask ourselves: why is the user typing ';' after hte namespcae name.  It's to use "file scoped namespaces".  in that case, we can just shortcircuit all the extra work they would need to do.  I can't really think of a scenario where we would not want to do that.

Reply:
As someone who types semicolons a lot, often accidentally, I definitely want this gated on that option. This is the same logic that was used to avoid having that option for ages "Why else would they type a semicolon?!".. well it turns out humans are fundamentally flawed :)

Reply:
I don't think it's much common that someone would type ; in this specific position. How often do someone modify a namespace line compared to writing actual code?